modules/renter/contractor/uptime.go

```diff
@@ -0,0 +1,76 @@
+package contractor
+
+import (
+	"time"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/modules"
+)
+
+// uptimeMinScans is the minimum number of scans required to judge whether a
+// host is offline or not.
+const uptimeMinScans = 3
+
+// uptimeWindow specifies the duration in which host uptime is checked.
+var uptimeWindow = func() time.Duration {
+	switch build.Release {
+	case "dev":
+		return 30 * time.Minute
+	case "standard":
+		return 7 * 24 * time.Hour // 1 week
+	case "testing":
+		return 15 * time.Second
+	}
+	panic("undefined uptimeWindow")
+}()
+
+// isOffline decides whether a host should be considered offline, based on its
+// scan metrics.
+func isOffline(host modules.HostDBEntry) bool {
+	// consider a host offline if:
+	// 1) The host has been scanned at least three times, and
+	// 2) The three most recent scans have all failed, and
+	// 3) The time between the most recent scan and the last successful scan
+	//    (or first scan) is at least uptimeWindow
+	numScans := len(host.ScanHistory)
+	if numScans < uptimeMinScans {
+		// not enough data to make a fair judgment
+		return false
+	}
+	// NOTE: ScanHistory is ordered from oldest-newest
```

Comment:
They adjust on the order of a few seconds, not days. It's only a problem if the discrepancy is enough to reorder the scans, which would mean two scans happened within seconds of each other.

Anyway, I thought of another potential issue: when determining the scan window, we call `windowEnd.Sub(windowStart)`. If the scans are not in order, it's possible for the subtraction to overflow.
In light of this, I am willing to sort the scans, despite my belief that such problems are highly unlikely to occur, and would be benign even if they did. I will also add some documentation and a debug check.